system/jhtree/jhinplace.cpp

```diff
@@ -0,0 +1,2681 @@
+{
+    if (keyLen != keyCompareLen)
+        lzwcomp.close();
+
+    hdr.keyBytes = getDataSize();
+
+    MemoryBuffer data;
+    data.setBuffer(maxBytes, keyPtr, false);
+    data.setWritePos(0);
+
+    if (positions.ordinality())
+    {
+        //Pack these by scaling and reducing the number of bytes
+        unsigned bytesPerPosition = 0;
+        if (minPosition != maxPosition)
+            bytesPerPosition = bytesRequired(maxPosition-minPosition);
+
+        bool useUncompressed = false;
+        byte sizeMask = (byte)bytesPerPosition | (scaleFposByNodeSize ? 0x80 : 0) | (useUncompressed ? 0x40 : 0);
+        data.append(sizeMask);
+        serializePacked(data, minPosition);
+        if (bytesPerPosition != 0)
+        {
+            for (unsigned i=0; i < positions.ordinality(); i++)
+            {
+                unsigned __int64 delta = positions.item(i) - minPosition;
+                serializeBytes(data, delta, bytesPerPosition);
+            }
+        }
+
+        if (isVariable)
+        {
+            //MORE: These may well benefit from packing....
+            ForEachItemIn(i, payloadLengths)
+                data.append((unsigned short)payloadLengths.item(i));
+        }
+
+        if (keyLen != keyCompareLen)
+        {
+            if (useUncompressed)
+            {
+                if (isVariable)
+                    serializePacked(data, uncompressed.length());
+                data.append(uncompressed.length(), uncompressed.bufferBase());
+            }
+            else if (keyLen != keyCompareLen)
+            {
+                size32_t payloadLen = lzwcomp.buflen();
+                serializePacked(data, payloadLen);
+                data.append(payloadLen, compressed.get());
+            }
+        }
+
+        builder.serialize(data);
+        ctx.numKeyedDuplicates += builder.numDuplicates;
+//        builder.gatherSingleTextCounts(ctx.singleCounts);
+
+        assertex(data.length() == getDataSize());
+    }
+
+    CWriteNode::write(out, crc);
+}
+
+
+//=========================================================================================================
+
+
+InplaceIndexCompressor::InplaceIndexCompressor(size32_t keyedSize, IHThorIndexWriteArg * helper)
+{
+    IOutputMetaData * format = helper->queryDiskRecordSize();
+    if (format)
+    {
+        //Create a representation of the null row - potentially used for the new compression algorithm
+        byte * nullRow = new byte[keyedSize];
+        RtlStaticRowBuilder rowBuilder(nullRow, keyedSize);
+
+        auto & meta = format->queryRecordAccessor(true);
+        size32_t offset = 0;
+        for (unsigned idx = 0; idx < meta.getNumFields() && offset < keyedSize; idx++)
+        {
+            const RtlFieldInfo *field = meta.queryField(idx);
+            offset = field->type->buildNull(rowBuilder, offset, field);
+        }
+        ctx.nullRow = nullRow;
+    }
+}
+
+InplaceIndexCompressor::~InplaceIndexCompressor()
+{
+    delete [] ctx.nullRow;
```

Comment:
Shouldn't ctc destructor be responsible for this?

Reply:
moved